relay-server/src/services/health_check.rs

```diff
@@ -138,3 +240,95 @@ impl Service for HealthCheckService {
         });
     }
 }
+
+#[derive(Debug)]
+struct SystemInfo {
+    system: System,
+    last_refresh: Instant,
+    refresh_interval: Duration,
+}
+
+impl SystemInfo {
+    pub fn new(refresh_interval: Duration) -> Self {
+        let system = System::new_with_specifics(
+            RefreshKind::new().with_memory(MemoryRefreshKind::everything()),
+        );
+
+        Self {
+            system,
+            last_refresh: Instant::now(),
+            refresh_interval,
+        }
+    }
+
+    pub fn memory(&mut self) -> Memory {
+        self.refresh();
+
+        // Use the cgroup if available in case Relay is running in a container.
+        if let Some(cgroup) = self.system.cgroup_limits() {
+            Memory {
+                used: cgroup.total_memory.saturating_sub(cgroup.free_memory),
```

Comment:
Not a blocker, but I wonder if we could base the health check on https://docs.rs/sysinfo/latest/sysinfo/struct.Process.html#method.memory, which returns the resident memory of the process.

This would require an additional indirection because we would need to configure what percentage of total memory our process is allowed to occupy. But it would prevent us from reporting unhealthy when Relay uses all the available memory for caching.

Reply:
We actually want to use system memory not process memory. It doesn't matter wether the container/host runs out of memory because Relay consumes all or something else is. In either case we want to stop accepting requests.